core/test.ts

```diff
@@ -0,0 +1,93 @@
+import { Session } from "@dataform/core/session";
+import * as table from "@dataform/core/table";
+import { dataform } from "@dataform/protos";
+
+export type TContextable<T> = T | ((ctx: TestContext) => T);
+
+export class Test {
+  public proto: dataform.ITest = dataform.Test.create();
+
+  public session: Session;
+  public contextableInputs: { [refName: string]: TContextable<string> } = {};
+
+  private datasetUnderTest: string;
+  private contextableQuery: TContextable<string>;
+
+  public dataset(datasetUnderTest: string) {
+    this.datasetUnderTest = datasetUnderTest;
+    return this;
+  }
+
+  public input(refName: string, contextableQuery: TContextable<string>) {
+    this.contextableInputs[refName] = contextableQuery;
+    return this;
+  }
+
+  public expect(contextableQuery: TContextable<string>) {
+    this.contextableQuery = contextableQuery;
+    return this;
+  }
+
+  public compile() {
+    if (!this.datasetUnderTest) {
+      this.session.compileError(new Error("Tests must operate upon a specified dataset."));
+      return;
+    }
+    const dataset = this.session.tables[this.datasetUnderTest];
+    if (!dataset) {
+      this.session.compileError(new Error(`Dataset ${this.datasetUnderTest} could not be found.`));
+      return;
+    }
+    if (dataset.proto.type === "incremental") {
+      this.session.compileError(
+        new Error(`Running tests on incremental datasets is not yet supported.`)
+      );
+      return;
+    }
+    const testContext = new TestContext(this);
+    const refReplacingContext = new RefReplacingContext(dataset, testContext);
+    this.proto.testQuery = refReplacingContext.apply(dataset.contextableQuery);
+    this.proto.expectedOutputQuery = testContext.apply(this.contextableQuery);
+    return this.proto;
+  }
+}
+
+export class TestContext {
+  public readonly test: Test;
+  constructor(test: Test) {
+    this.test = test;
+  }
+
+  public apply<T>(value: TContextable<T>): T {
+    if (typeof value === "function") {
+      return (value as any)(this);
+    } else {
+      return value;
+    }
+  }
+}
+
+class RefReplacingContext extends table.TableContext {
```

Comment:
probably shouldn't extend here - as we still have those old methods around in TableContext that could have side-effects. Unfortunately, I think we need an ITableContext interface and to explicitly implement no-ops for all the other methods.